### ISSUE TITLE
Using skip() of tf.data.Dataset is slow for big datasets

### ISSUE THREAD

**Issue by benkli01** [tensorflow/tensorflow#44008]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.6.9
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: -

**Describe the current behavior**

When using _skip()_ to skip most of a big _tf.data.Dataset_ (e.g. to create a validation set from a training split by just taking the last 10% of the training data), iteration over the dataset takes very long to begin. In the standalone code below skipping samples takes about 1.5 minutes to return the first sample. Another test with imagenet took ~25 minutes before returning the first sample.

**Describe the expected behavior**

1. I would expect the first sample from a dataset to be available in a shorter time (seconds not minutes, <10 seconds). 
2. I would also expect the time to get the first sample when skipping to depend on the size of the shards not on the size of the dataset.

**Standalone code to reproduce the issue**

Below is a small script using COCO from tfds to demonstrate the problem.

For me the output is something like this:
> First sample for complete set: 0.49 sec
> First sample for skipped set: 89.66 sec
> 

Standalone code:
```python
import tensorflow_datasets as tfds
import time

def test_tfds(dataset_name='coco'):
    print('### Test tfds ###')

    # Download dataset
    print("Downloading dataset '{}'...".format(dataset_name))
    builder = tfds.builder(dataset_name)
    download_dir = dataset_name
    builder.download_and_prepare(download_dir=download_dir)
    print('...dataset downloaded:\n{}'.format(builder.info.splits))

    # Get train split 
    split = 'train'
    dataset = tfds.load(dataset_name, split=split)
    
    test(dataset, builder.info.splits[split].num_examples)

def test(dataset, num_samples):
    # Get first sample of complete dataset: this should be ok
    t = time.time()
    _ = next(iter(dataset))
    print('First sample for complete set: {:.2f} sec'.format(time.time() - t))

    # Now skip 90% of the train split
    num_skip_samples = int(0.9 * num_samples)
    skipped_dataset = dataset.skip(num_skip_samples)

    # Get first sample of skipped dataset: this takes extremely long
    t = time.time()
    _ = next(iter(skipped_dataset))
    print('First sample for skipped set: {:.2f} sec'.format(time.time() - t))

if __name__ == '__main__':
    test_tfds('coco')
```

**Comment by Saduf2019** [comment]
@benkli01 
I ran the code shared , please find the [gist here](https://colab.research.google.com/gist/Saduf2019/80c2269b05f0024295dcf2154977dd14/untitled439.ipynb),and let us know it confirms the issue reported.

**Comment by benkli01** [comment]
Thank you @Saduf2019 . The final output with the timing information seems to be missing. Did the process finish?

**Comment by Saduf2019** [comment]
@benkli01 
Please share "/Python/Datasets/v1.0/train/" this file for us to replicate.

**Comment by benkli01** [comment]
@Saduf2019 I'm afraid I don't understand what file you mean? Could you elaborate please? There is no such a file ("/Python/Datasets/v1.0/train/") on my machine. The example should be self-contained and download everything it needs.

**Comment by jsimsa** [comment]
The tf.data `skip(n)` transformation does not avoid the computation needed to process the `n` elements, it just discards them. This is because in general, tf.data sources and transformations do not provide an efficient skip implementation (e.g. a compressed file without an index needs to be fully decompressed and iterated to determine how many elements it contains or the `filter` transformation needs to evaluate all inputs in order to determine which of them to keep).

If you would like to efficiently divide your dataset into multiple "shards", you should do so at the file-level.

**Comment by jsimsa** [comment]
Sorry, didn't mean to close the issue. Feel free to close it, if my answer provides satisfactory explanation for why `skip` works the way it does.

**Comment by benkli01** [comment]
Hi @jsimsa , 
thank you for your explanation. I was suspecting something like this. 
So the general process for skipping is 1) consuming/loading samples and 2) discarding them immediately?

Could this be implemented more efficiently for specific types of datasets to speed things up? E.g. the COCO dataset, as used in my example script, is just saved as a bunch of jpg files on the disk (with one file equal to one sample) that could be easily skipped without loading the files.

**Comment by jsimsa** [comment]
If you were to use tf.data directly (as opposed to through TFDS), you could use the `skip` transformation to filter filenames to read from, which will provide you with an efficient implementation.

Since you are using TFDS, you cannot easily do that because TFDS abstracts away the details it uses for loading the dataset. Having said that, I believe that you should be able to use the [Split](https://github.com/tensorflow/datasets/blob/master/docs/splits.md) API to inform TFDS what subset of elements would you like to extract and TFDS should be able to translate this to an efficient implementation (i.e. skipping at the level of files as opposed to loading everything and dropping a subset).

**Comment by benkli01** [comment]
Thanks @jsimsa . Using the split API when loading the dataset speeds things up significantly compared to skipping samples afterwards. Unfortunately from what I've gathered this doesn't really work for TFRecords though as they do not encode the number of samples per file, so that the iteration over all samples is unavoidable. Please correct me, if I'm wrong.
I'm closing this ticket now. Thanks again for the answers!

**Comment by jsimsa** [comment]
To enable efficient skipping of data for TFRecords (or any other format that does not support efficient indexing), you can be provide coarse-grained indexing functionality yourself by sharding your data across multiple files (and either using a fix number of elements per file or storing the number of files in filename).
